presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSourceProvider.java

```diff
@@ -648,13 +649,18 @@ public ConnectorPageSource createPageSource(
             ConnectorSession session,
             ConnectorSplit connectorSplit,
             ConnectorTableLayoutHandle layout,
-            List<ColumnHandle> columns,
+            List<ColumnHandle> desiredColumns,
             SplitContext splitContext)
     {
         IcebergSplit split = (IcebergSplit) connectorSplit;
         IcebergTableLayoutHandle icebergLayout = (IcebergTableLayoutHandle) layout;
         IcebergTableHandle table = icebergLayout.getTable();
 
+        List<ColumnHandle> columns = desiredColumns;
+        if (split.getChangelogSplitInfo().isPresent()) {
+            columns = (List<ColumnHandle>) (List<?>) split.getChangelogSplitInfo().get().getIcebergColumns();
```

Comment:
So the `desiredColumns` field represents the list of columns the query is interested in reading. i.e. if you had table `t(c1 int, c2 int)`, and queried `SELECT c1 FROM t` this list would only contain the columns from `c1`.

In the case of the changelog, if we queried the changelog with

```sql
SELECT ordinal FROM t$changelog
```

Then the `desiredColumns` list is populated with the handle to the `ordinal` column. However, that column doesn't actually exist in the parquet file. We need to pass a valid set of columns to the `createPageDataSource` function to build a parquet/ORC reader that only reads the desired columns. In the case of the changelog, we will always read the full set.

In summary, This conditional exists because we need to pass the "real" column set from the source table schema into the `createPageDataSource`, rather than the "virtual" columns from the changelog schema to properly read the source data.